refnx/reduce/reduce.py

```diff
@@ -558,6 +559,439 @@ def _reduce_single_angle(self, scale=1):
+        ):
+            self.reducers["du"].reflected_beam = self.reducers[
+                "ud"
+            ].reflected_beam
+        elif hasattr(self.reducers["du"], "reflected_beam") and not hasattr(
+            self.reducers["ud"], "reflected_beam"
+        ):
+            self.reducers["ud"].reflected_beam = self.reducers[
+                "du"
+            ].reflected_beam
+        elif not hasattr(
+            self.reducers["du"], "reflected_beam"
+        ) and not hasattr(self.reducers["ud"], "reflected_beam"):
+            self.reducers["ud"].reflected_beam = np.zeros(np.shape(wl))
+            self.reducers["du"].reflected_beam = np.zeros(np.shape(wl))
+
+        # Check whether mode is POLANAL or just POL instead of this
+        if self.reducers["dd"].reflected_beam.cat.mode == "POL":
+            # if out of the beam, polarisation config only uses
+            # polariser and flipper1
+            config = "PF"
+        elif self.reducers["dd"].reflected_beam.cat.mode == "POLANAL":
+            # if in the beam, polarisation config uses all elements.
+            config = "full"
+
+        if poleff is None:
+            # Define polarisation efficiency of PLATYPUS
+            poleff = PolarisationEfficiency(
+                self.reducers["dd"].reflected_beam.m_lambda, config=config
+            )
+
+        # Define sizes of corrected beam spectra (T,S,N) and
+        # inverted combined efficiency matrix to be (T,4,4)
+        # where T is the number of wavelength bins
+
+        # Create dict of direct and reflected beam spectra
+        rb_spectra = {}
+        db_spectra = {}
+        for sc in self.reducers:
+            try:
+                # Try assigning reflected beam spectra via reducer
+                rb_spectra[sc] = self.reducers[sc].reflected_beam.m_spec
+                db_spectra[sc] = self.reducers[sc].direct_beam.m_spec
+
+                if sc in ["du", "ud"]:
+                    # Need to get spin-flip direct beams from SpinSet since
+                    # we don't include them in the reducer
+                    db_spectra[sc] = getattr(
+                        self.spin_set_direct, f"{sc}"
+                    ).m_spec
+
+            except AttributeError:
+                print(f"{sc} not measured")
+                # if spin channel not measured, assumed
+                # intensity values are zero
+                db_spectra[sc] = np.zeros(np.shape(wl))
+                rb_spectra[sc] = np.zeros(np.shape(wl))
+
+        # Invert and apply the refnx.reduce.PolarisationEfficiency parameters
+        # to the raw spectra to correct for efficiencies.
+        inverted_combined_efficiency_matrix = np.linalg.inv(
+            poleff.combined_efficiency_matrix
+        )
+        # Create numpy arrays with shape (T,S,N) (Time-of-flight,
+        # spin-channel, dataset_number) to broadcast with array of
+        # efficiency matrices
+
+        raw_db = np.empty((len(wl[0]), 4, 1))
```

Comment:
If the efficiency matrix is (T, 4, 4), then I think raw_db should be (N, T, 4)
We want something along the lines of:

(1, T, 4, 4) @ (N, T, 4)  -> (N, T, 4)

Reply:
```
>>> A = np.random.random(size=(3, 2, 4, 4))
>>> B = np.random.random(size=(3, 2, 4))

>>> (A @ B[..., None])[0]
array([[[1.00677652],
        [0.68032058],
        [0.98081147],
        [0.63686509]],

       [[2.02502462],
        [1.54192624],
        [1.33315182],
        [0.90808536]]])
>>> (A @ B[..., None]).shape
(3, 2, 4, 1)

>>> A[0, 0] @ B[0, 0] # (4, 4)
array([1.00677652, 0.68032058, 0.98081147, 0.63686509])
```